### PR TITLE
feat(pkg/objstore/s3): support for prefixes

### PR DESCRIPTION
**Pull request by jaredallard**
* [] CHANGELOG Support for s3 prefixes

## Changes

I added support for s3 prefixes based on the amazing work proposed by #1318 

## Verification

Tested configuration changes via unit test, tested integration wise by running it.

### CODE REVIEW

**Comment by jaredallard** [comment]
Going to fix the docs related stuff... as soon as I figure out how that works

**Comment by jaredallard** [comment]
Ready to go :tada:

File: pkg/objstore/s3/s3.go [hunk 1]
```diff
@@ -92,6 +94,11 @@ func parseConfig(conf []byte) (Config, error) {
 		config.PartSize = defaultMinPartSize
 	}
 
+	// here for testing
+	if config.Prefix != "" {
+		config.Prefix = strings.TrimSuffix(config.Prefix, DirDelim) + DirDelim
```
Comment (povilasv) [inline pkg/objstore/s3/s3.go]: Why is this needed?

File: pkg/objstore/s3/s3.go [hunk 2]
```diff
@@ -208,6 +216,17 @@ func validate(conf Config) error {
 	if conf.AccessKey != "" && conf.SecretKey == "" {
 		return errors.New("no s3 secret_key specified while access_key is present in config file; either both should be present in config or envvars/IAM should be used.")
 	}
+
+	// ensure that the prefix isn't too long
+	if conf.Prefix != "" && len(conf.Prefix) >= 1024 {
+		return errors.New("prefix is too long (limited by Amazon at 1024 bytes long)")
```
Comment (povilasv) [inline pkg/objstore/s3/s3.go]: I don't think this is a good idea, as S3 isn't only amazon it's also minio, ceph and other s3 compatible stores, so if other implementations support long prefixes, this would break those.

File: pkg/objstore/s3/s3.go [hunk 3]
```diff
@@ -208,6 +216,17 @@ func validate(conf Config) error {
...
+	}
+
+	// trim trailing slash and insert one
```
Comment (povilasv) [inline pkg/objstore/s3/s3.go]: Why are we doing this?

File: pkg/objstore/s3/s3.go [hunk 4]
```diff
@@ -230,16 +249,17 @@ func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) err
 		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
 	}
 
-	for object := range b.client.ListObjects(b.name, dir, false, ctx.Done()) {
+	for object := range b.client.ListObjects(b.name, b.objectPrefix+dir, false, ctx.Done()) {
+		keyName := strings.Replace(object.Key, b.objectPrefix, "", 1)
```
Comment (povilasv) [inline pkg/objstore/s3/s3.go]:
```suggestion
		keyName := strings.TrimPrefix(object.Key, b.objectPrefix)
```

File: pkg/objstore/s3/s3.go (hunk 1 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: for the configuration parsing test

File: pkg/objstore/s3/s3.go (hunk 3 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: normalization of paths for object keys, don't want two `/` on accident.

File: pkg/objstore/s3/s3.go (hunk 4 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: Good one, I should've caught that.

File: pkg/objstore/s3/s3.go (hunk 2 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: OK, I can remove that.

File: pkg/objstore/s3/s3.go (hunk 1 above)
Comment (povilasv) [reply pkg/objstore/s3/s3.go]: Adding normal code just for tests doesn't sound like a good idea, consider removing this.

File: pkg/objstore/s3/s3.go (hunk 3 above)
Comment (povilasv) [reply pkg/objstore/s3/s3.go]: Consider adding user validation, instead of trying to fix the error.

File: pkg/objstore/s3/s3.go (hunk 1 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: Alright.

File: pkg/objstore/s3/s3.go (hunk 3 above)
Comment (jaredallard) [reply pkg/objstore/s3/s3.go]: Not sure I agree with that, I think this is a perfectly valid solution,

**Comment by jaredallard** [comment]
@edevouge feel free to nullify this PR with your work, I needed to implement this for my own use cases, and I would much rather you get credit than I.

**Comment by bwplotka** [comment]
So, it looks good, but why are we limiting ourselves to S3 implementation? If we are getting into this, I think it might be a part of our interface instead. E.g Adding `Prefix` to https://github.com/thanos-io/thanos/blob/2c5f2cde11f5cd100f147ad2e5d4dbeccbd604c5/pkg/objstore/client/factory.go#L32 

This will be then consistent across all object storages.

**Comment by edevouge** [comment]
> @edevouge feel free to nullify this PR with your work, I needed to implement this for my own use cases, and I would much rather you get credit than I.

@jaredallard : I appreciate your contribution, but it would be great the next time you decided to implement a design proposal to post a small comment on the issue thread to explain your intention. Because, I was also writing some code for this feature and it's not really efficient to do the job twice. I'll let you finish the job.

**Comment by jaredallard** [comment]
> So, it looks good, but why are we limiting ourselves to S3 implementation? If we are getting into this, I think it might be a part of our interface instead. E.g Adding `Prefix` to
> 
> https://github.com/thanos-io/thanos/blob/2c5f2cde11f5cd100f147ad2e5d4dbeccbd604c5/pkg/objstore/client/factory.go#L32
> 
> This will be then consistent across all object storages.

Good point, I'll rework this.

**Comment by AnandDevan** [comment]
hi folks, is this waiting for the equivalent implementation on the other object stores for merging?
would like to see prefix supported for s3 soon

**Comment by svyotov** [comment]
This looks like a good feature, do you think it will re-open?

File: pkg/objstore/s3/s3.go [hunk 5]
```diff
@@ -87,6 +89,11 @@ func parseConfig(conf []byte) (Config, error) {
 		return Config{}, err
 	}
 
+	// here for testing
```
Comment (bwplotka) [inline pkg/objstore/s3/s3.go:92]: This looks wrong as @povilasv mentioned in https://github.com/thanos-io/thanos/pull/1392/files#r312902398

File: docs/storage.md [hunk 6]
```diff
@@ -90,6 +90,7 @@ config:
     insecure_skip_verify: false
   trace:
     enable: false
+  prefix: ""
```
Comment (bwplotka) [inline docs/storage.md:93]: I would be curious and actually I think it's must-have to actually make this change on bucket level, not each bucket client level.. WDYT? (:

**Comment by bwplotka** [comment]
Well it really depends on @jaredallard if he is willing to continue this change.

I believe to merge it we need to do it per global object storage config, not just S3, so all providers can reuse this prefix logic (: We can also discuss if it should be `Prefix` or `Prefixes` potentially on issue: https://github.com/thanos-io/thanos/issues/1318

Help wanted, we definitely would like to have this in!

**Comment by jaredallard** [comment]
Hey, I'm likely to not drive this work to completion. I am no longer on the team that needed this change at my company (they have maintained a fork with this AFAIK). I may come and do it in my spare time, but I have a massive backlog of things to do outside of work and this is pretty low on that list for that reason 😓 

Sorry everyone!

**Comment by bwplotka** [comment]
This is totally ok (: No worries, and good luck! 

It's best if you are transparent about the priorities, so someone else can take your work. :+1 Thanks!


so... let's keep it close and help wanted (:
